### PR TITLE
Refactor Optimizer class

### PR DESCRIPTION
**Pull request by guillaumekln**
The main change is to serialize the optimizer state and not the instance itself. Saving the instance restricted changes applied to the class and the use of optimizer wrappers. This change is backward compatible with existing checkpoints.

The learning rate management is also updated: the learning rate value is always computed dynamically based on the initial learning rate and the current training step. The learning rate decay schedules are expected to be passed as callable taking a step and returning a scaling factor.

### CODE REVIEW

File: onmt/utils/optimizers.py [hunk 1]
```diff
@@ -2,87 +2,108 @@
...
-                optim.max_grad_norm = opt.max_grad_norm
-                optim.lr_decay = opt.learning_rate_decay
-                optim.start_decay_steps = opt.start_decay_steps
-                optim.decay_steps = opt.decay_steps
-                optim.betas = [opt.adam_beta1, opt.adam_beta2]
-                optim.adagrad_accum = opt.adagrad_accumulator_init
-                optim.decay_method = opt.decay_method
-                optim.warmup_steps = opt.warmup_steps
-                optim.model_size = opt.rnn_size
+def make_optimizer(model, opt):
```
Comment (bpopeters) [inline onmt/utils/optimizers.py]: Why the change to the function name? I am open to all sorts of changes with how objects are created, but all the other ones start with 'build' and this breaks that consistency.

File: onmt/utils/optimizers.py [hunk 2]
```diff
@@ -132,119 +153,123 @@ class Optimizer(object):
...
+        """
+        self._optimizer = optimizer
+        self._learning_rate = learning_rate
+        self._learning_rate_decay_fn = learning_rate_decay_fn
+        self._max_grad_norm = max_grad_norm
+        self._training_step = 1
+        self._decay_step = 1
+
+    @staticmethod
+    def from_options(model, opt, checkpoint=None):
```
Comment (bpopeters) [inline onmt/utils/optimizers.py]: I really like the idea of making objects with something like this instead of the current proliferation of functions whose names start with 'build' or 'make'. Would it make sense, though, for this to be a class method? It is defined on the class `Optimizer` and returns an `Optimizer` instance. Plus, a class method would might make it easier if someone wanted to subclass `Optimizer` later.

File: onmt/utils/optimizers.py (hunk 1 above)
Comment (guillaumekln) [reply onmt/utils/optimizers.py]: This is a new function that creates a `torch.optim.Optimizer` instance. The previous `build_optim` function has been moved to `Optimizer.from_options` to avoid confusion.

Maybe we should rename the `Optimizer` class to something else to avoid confusion with the PyTorch class with the same name?

File: onmt/utils/optimizers.py (hunk 2 above)
Comment (guillaumekln) [reply onmt/utils/optimizers.py]: I think this should be a static function. Otherwise you have to create an instance with an incomplete state.

File: onmt/utils/optimizers.py [hunk 3]
```diff
@@ -2,87 +2,108 @@
...
+        model_size ** (-0.5) *
+        min(step ** (-0.5),
+            step * warmup_steps**(-1.5)))
+
+
+def exponential_decay_fn(rate, decay_steps, start_step=0):
+    """A standard exponential decay, scaling the learning rate by :obj:`rate`
+    every :obj:`decay_steps` steps.
+    """
+    return lambda step: (
```
Comment (bpopeters) [inline onmt/utils/optimizers.py]: Is there a case where an `Optimizer` object might need to be serialized? If so, I think returning a lambda here might be a problem because this lambda expression will end up as an attribute of the `Optimizer`. The only advantage lambdas have over other ways of defining functions is that they can give you short one-liners; other than that, they have only disadvantages. The code here looks like it could be rewritten with `functools.partial`, which is serializable.

Comment (bpopeters) [reply onmt/utils/optimizers.py]: Or just with normal functions.

File: onmt/utils/optimizers.py (hunk 2 above)
Comment (bpopeters) [reply onmt/utils/optimizers.py]: How so?

Comment (guillaumekln) [reply onmt/utils/optimizers.py]: I mean if it's a method you first need an instance, but you haven't yet looked at the options to know how to create it.

Comment (bpopeters) [reply onmt/utils/optimizers.py]: Class methods (with the decorator `@classmethod`) do not require an instance. The first argument is conventionally `cls` instead of `self` and it is a reference to the class, not an instance of the class. This looks like a good place to use it because the method _does_ use the class. But it does it by hard-coding it in, which would not be necessary with a class method.

Comment (guillaumekln) [reply onmt/utils/optimizers.py]: Oh sorry I was confused by the term "class method" at first. Yes, that makes sense!

File: onmt/utils/optimizers.py (hunk 1 above)
Comment (vince62s) [reply onmt/utils/optimizers.py]: I agree the new structure make things unclear.
IMO:
let's keep build_optim to build the pytorch Optimizer
let's rename our class Optimizer to Optim_wrapper
I don't  like the from_options name. Would the call Optim_wrapper.build() make sense ?
open to other ideas.

Comment (bpopeters) [reply onmt/utils/optimizers.py]: I don't much like the `OptimWrapper` name because it's presenting things at a level of abstraction that is not useful to people who work on machine translation. Optimizers I understand. Wrappers, managers, builders -- things like this are not so interesting to me and make my life harder.

I like the name `from_options` (or maybe `from_opt` if it's clearer). It is a pattern for constructor naming that is well established elsewhere (numpy arrays have `fromfile`, `fromfunction`, `fromiter`, etc.; pandas dataframes have `from_dict`, `from_items`, etc.). I think this is much better than a name like `build()`, which is completely uninformative.

Comment (vince62s) [reply onmt/utils/optimizers.py]: I am not fighting this. But if so, there are plenty of places where we would need to rename to make things consistent all the way through. So far, we used build to create instances.
As for the Optimizer there is definitely an overlap with the Pytorch native optimizer class. Zero abstraction in names when there is, is not good either.

Comment (vince62s) [reply onmt/utils/optimizers.py]: That won't happen soon if we keep 1) Adafactor 2) SparseAdam (which is not the same as the legacy pytorch SparseAdam Optimizer that supports onyl sparse tensors). We could drop these 2, but then it's another story. In the meantime, I'm fine too leave things as they are, not optimal but ok.

Comment (bpopeters) [reply onmt/utils/optimizers.py]: I don't follow. Is there a reason why our own implementations of Adafactor and SparseAdam are incompatible with `torch.optim.lr_scheduler`? I see that Adafactor already subclasses `torch.optim.Optimizer`. I don't see why it wouldn't be compatible with `lr_scheduler` as long as it uses the same interface as the optimizers in `torch.optim`, which also subclass `torch.optim.Optimizer`.

Comment (guillaumekln) [reply onmt/utils/optimizers.py]: I did consider the LR scheduling from PyTorch but did not found them to add much value over a simple and flexible function based approach. And they are using the epoch terminology which would be confusing in our case.

Regarding Adafactor, the issue I think is that the LR scheduling is defined in the optimizer. It should not be, and it is now very easy to add an additional decay method. I can add that as part of this PR.

Regarding the naming discussed above, I propose:

* `make_optimizer` -> `build_torch_optimizer`
* `Optimizer.from_options` -> `Optimizer.from_opt`

**Comment by vince62s** [comment]
thanks, merging.
